clientconn.go

```diff
@@ -1231,8 +1228,7 @@ func (ac *addrConn) adjustParams(r transport.GoAwayReason) {
 	}
 }
 
-func (ac *addrConn) resetTransport() {
-	ac.mu.Lock()
+func (ac *addrConn) resetTransportAndUnlock() {
```

Comment:
Please add a short comment here:
```go
// resetTransportAndUnlock unconditionally connects the addrConn.
//
// ac.mu must be held by the caller, and this function will guarantee it is released.
```

Reply:
> ac.mu must be held by the caller, and this function will guarantee it is released

should we have code check for this as well?

Reply:
Add the doc comment.

Reply:
> should we have code check for this as well?
We don't verify if the mutex is locked in other functions in the code which assume the caller has the lock. These functions have the suffix `Locked` in the name. 
* There is a `TryLock()` method on `sync/mutex`, but its use is discouraged. 
* This is a private method and we run tests with race detector so we can catch incorrect usage.
* When the `resetTransportAndUnlock` method tries to unlock a mutex that isn't locked, it will panic.

Reply:
i meant just having a doc doesn't enforce the mutex is locked. Unlocking a mutex that is not locked in Go will result in a runtime panic.

Reply:
Can you suggest how to enforce the locking? 
How is the caller expected to handle the error if `resetTransportAndUnlock` doesn't panic?
Since its a private method, doesn't a panic make the failure more visible and ensure incorrect usages are caught by tests?

Reply:
yeah it will probably require to implement custom mutex with some boolean field but since we have precedences of methods (e.g https://github.com/grpc/grpc-go/blob/master/clientconn.go#L728) unlocking mutex, devs will be aware of these

Reply:
I think the name of the function and the comment should be sufficient for this.